**Asynchronous Programming**

**Easy**
1. What is asynchronous programming?
2. Difference between synchronous and asynchronous execution.
3. What is an event loop?
4. What is a coroutine?
5. What is await?

**Medium**

6. Difference between asyncio.gather() and sequential await.
7. Difference between threading and asyncio.
8. When should you use multiprocessing?
9. What happens if one coroutine fails in asyncio.gather()?
10. What is asyncio.create_task()?

**Hard**

11. Build an async web scraper.
12. Download 100 URLs concurrently.
13. Limit concurrency using Semaphore.
14. Retry failed API calls asynchronously.
15. Design an async task queue.

**1. What is asynchronous programming?**

**Definition**

Asynchronous programming is a `non-blocking programming technique` that allows `multiple operations` to `progress independently`, `improving responsiveness` and `resource utilization`.

**Example:**

Suppose an application needs to download a file from the internet.

- **Synchronous:** The program waits until the download is complete before doing anything else.
  
- **Asynchronous:** The program starts the download, continues performing other work, and processes the file once the download is finished.


**Key Features**

- **Non-blocking execution** – the program does not pause while waiting for long-running operations.
- **Concurrency** – multiple tasks can make progress during the same period.
- **Better responsiveness** – applications remain responsive to users.
- **Efficient resource utilization** – CPU time is not wasted waiting for I/O operations.

**Common Applications**

- Web applications and APIs
- Database operations
- File input/output (I/O)
- Network communication
- Real-time messaging and streaming

**2. Difference between synchronous and asynchronous execution.**

| **Synchronous Execution**                                     | **Asynchronous Execution**                                                    |
| ------------------------------------------------------------- | ----------------------------------------------------------------------------- |
| Tasks are executed one after another.                         | Tasks can start without waiting for previous tasks to finish.                 |
| It is **blocking**; the program waits until a task completes. | It is **non-blocking**; the program continues with other tasks while waiting. |
| Simpler to write and understand.                              | More complex due to callbacks, promises, or async/await.                      |
| Slower for I/O-intensive operations because of waiting.       | Faster and more efficient for I/O-intensive operations.                       |
| Suitable for small or simple programs.                        | Suitable for web servers, network applications, and real-time systems.        |
| Less responsive during long-running tasks.                    | More responsive, as the application remains active while tasks run.           |

**Example**

**Synchronous Execution**

```
Start
↓
Read File (wait)
↓
Process Data
↓
End

```

**Asynchronous Execution**

```
Start
↓
Read File (background)
↓
Do Other Work
↓
Process File When Ready
↓
End
```

**What is an Event Loop?**

**Definition**

An event loop is a program mechanism that waits for events, processes them, and executes associated tasks, enabling asynchronous and non-blocking programming.


**How the Event Loop Works**

1. Executes synchronous code first.
2. Registers asynchronous operations (e.g., timers, file I/O, network requests).
3. Continues running without waiting for those operations to complete.
4. When an asynchronous operation finishes, its callback or task is placed in a queue.
5. The event loop takes tasks from the queue and executes them when the call stack is empty.

**Simple Flow**

```
Start
   ↓
Execute Synchronous Code
   ↓
Start Asynchronous Task
   ↓
Continue Other Work
   ↓
Task Completes
   ↓
Event Loop Picks Callback
   ↓
Execute Callback
   ↓
End
```

**Advantages**

- Enables non-blocking execution.
- Keeps applications responsive.
- Efficiently handles multiple asynchronous tasks.
- Widely used in JavaScript, Node.js, Python (asyncio), and other asynchronous runtimes.

**4. What is a Coroutine?**

**Definition:**

A coroutine is a function that can suspend (pause) and resume its execution, allowing efficient handling of asynchronous tasks without blocking the program.

**Key Features**

- Can pause execution using `await` (or similar keywords).
- Resumes execution when the awaited task is complete.
- Supports non-blocking operations.
- Makes asynchronous code easier to read and maintain.

**Example**

In [13]:
import asyncio

async def greet():
    print("Hello")
    await asyncio.sleep(2)   # Pause for 2 seconds
    print("Welcome!")

await greet()

Hello
Welcome!


**How It Works**

1. A coroutine starts executing.
2. It reaches an await statement and pauses.
3. The event loop executes other tasks while waiting.
4. When the awaited operation completes, the coroutine resumes from the point where it paused.

**Advantages**

- Simplifies asynchronous programming.
- Improves application responsiveness.
- Efficiently handles I/O-bound operations.
- Avoids blocking the main thread.

**Common Uses**

- Network requests
- Database operations
- File I/O
- Web servers
- Real-time applications

**What is `await?`**

`await` is a keyword that suspends the execution of a coroutine until the awaited asynchronous task finishes, without blocking the program.

**Key Features**

- Used only inside an async function or coroutine.
- Waits for an asynchronous operation to complete.
- Does not block the entire program.
- Improves code readability by making asynchronous code look similar to synchronous code.

**Example**

In [20]:
import asyncio

async def fetch_data():
    print("Fetching data...")
    await asyncio.sleep(2)   # Wait asynchronously
    print("Data received")

await fetch_data()

Fetching data...
Data received


**How await Works**

1. The coroutine starts executing.
2. When it reaches await, it pauses.
3. The event loop runs other tasks while waiting.
4. Once the awaited operation completes, the coroutine resumes from where it paused.

**Advantages**
- Prevents blocking the main thread.
- Makes asynchronous code cleaner and easier to understand.
- Improves application responsiveness.
- Efficiently handles I/O-bound tasks.

## **Medium**

**6. Difference between asyncio.gather() and sequential await.**

| **`asyncio.gather()`**                                                        | **Sequential `await`**                                         |
| ----------------------------------------------------------------------------- | -------------------------------------------------------------- |
| Runs multiple coroutines **concurrently**.                                    | Runs coroutines **one after another**.                         |
| Total execution time is usually shorter for independent I/O-bound tasks.      | Total execution time is the sum of all task durations.         |
| Suitable when tasks are independent of each other.                            | Suitable when each task depends on the previous task's result. |
| Returns the results of all coroutines together in the order they were passed. | Returns each result individually as each `await` completes.    |
| Improves performance for concurrent operations.                               | Simpler when tasks must execute in sequence.                   |


In [25]:
import asyncio

async def task1():
    await asyncio.sleep(2)
    return "Task 1"

async def task2():
    await asyncio.sleep(2)
    return "Task 2"

async def main():
    result1 = await task1()
    result2 = await task2()
    print(result1, result2)

await main()

Task 1 Task 2


Execution time: About 4 seconds (2 + 2).

**Example: `asyncio.gather()`**

In [28]:
import asyncio

async def task1():
    await asyncio.sleep(2)
    return "Task 1"

async def task2():
    await asyncio.sleep(2)
    return "Task 2"

async def main():
    result1, result2 = await asyncio.gather(task1(), task2())
    print(result1, result2)

await main()

Task 1 Task 2


Execution time: About 2 seconds, because both tasks run concurrently.

**When to Use**

- Use asyncio.gather() when:
  - Tasks are independent.
  - You want better performance by running them concurrently.
  - You need all results together.
- Use sequential await when:
  - One task depends on the output of another.
  - The tasks must execute in a specific order.

**7. Difference between threading and asyncio.**

| **Threading**                                                                                       | **`asyncio`**                                                                                          |
| --------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------ |
| Uses **multiple threads** to execute tasks.                                                         | Uses a **single thread** with an event loop to run asynchronous tasks.                                 |
| Suitable for **I/O-bound** tasks and some concurrent operations.                                    | Best suited for **I/O-bound** and non-blocking asynchronous tasks.                                     |
| Threads can run independently and may execute in parallel (depending on the runtime and task type). | Coroutines run cooperatively; only one coroutine executes at a time on the event loop.                 |
| Requires synchronization mechanisms (locks, semaphores, etc.) when sharing data.                    | Usually avoids shared-memory synchronization because tasks cooperate by yielding control with `await`. |
| Higher memory and context-switch overhead.                                                          | Lower memory usage and minimal context-switch overhead.                                                |
| Uses the `threading` module.                                                                        | Uses the `asyncio` module with `async` and `await`.                                                    |
| Better when working with blocking libraries that cannot be made asynchronous.                       | Better when using asynchronous libraries (networking, databases, file I/O).                            |


**Example: Threading**

In [36]:
import threading
import time

def task():
    print("Task started")
    time.sleep(2)
    print("Task completed")

t = threading.Thread(target=task)
t.start()
t.join()

Task started
Task completed


**Example: `asyncio`**

In [39]:
import asyncio

async def task():
    print("Task started")
    await asyncio.sleep(2)
    print("Task completed")

await task()

Task started
Task completed


**When to Use**

**Use Threading when:**

- You are working with blocking code or libraries.
- You need to perform background tasks alongside the main program.
- Converting the code to asynchronous style is not practical.

**Use asyncio when:**

- You have many I/O-bound operations (network requests, APIs, databases).
- You want scalable, non-blocking applications.
- You are using libraries that support async/await.

**8. When should you use multiprocessing?**

**Definition:**

Multiprocessing is a programming technique that creates multiple processes to execute tasks in parallel, **improving performance for CPU-intensive applications.**

**When to Use Multiprocessing**

- CPU-intensive calculations (e.g., mathematical computations, simulations).
- Image and video processing.
- Machine learning and data analysis.
- Scientific computing.
- Large-scale data processing.
- Tasks that need to fully utilize multiple CPU cores.

**Example (Python)**

In [46]:
from multiprocessing import Process
import time

def task():
    print("Task started")
    time.sleep(2)
    print("Task completed")

if __name__ == "__main__":
    p1 = Process(target=task)
    p2 = Process(target=task)

    p1.start()
    p2.start()

    p1.join()
    p2.join()

**Advantages**

- True parallel execution on multiple CPU cores.
- Faster execution for CPU-bound tasks.
- Each process has its own memory space, improving isolation.
- Avoids Python's Global Interpreter Lock (GIL) for CPU-bound work.

**Disadvantages**
- Higher memory usage than threads.
- Creating and managing processes has more overhead.
- Communication between processes is slower and more complex.

**Multiprocessing vs Threading vs Asyncio**

| Feature     | Multiprocessing                  | Threading                                      | Asyncio                       |
| ----------- | -------------------------------- | ---------------------------------------------- | ----------------------------- |
| Execution   | Multiple processes               | Multiple threads                               | Single thread with event loop |
| Best for    | CPU-bound tasks                  | Blocking I/O tasks                             | Non-blocking I/O tasks        |
| Parallelism | Yes                              | Limited for CPU-bound Python code (due to GIL) | No (concurrent, not parallel) |
| Memory      | Separate memory for each process | Shared memory                                  | Shared memory in one thread   |
| Overhead    | High                             | Medium                                         | Low                           |


**9. What happens if one coroutine fails in asyncio.gather()?**

By default, if one coroutine raises an exception, `asyncio.gather()` immediately raises that exception to the caller. The other coroutines are not guaranteed to complete normally; pending tasks may be cancelled depending on the situation.

**Example**

In [52]:
import asyncio

async def task1():
    await asyncio.sleep(1)
    return "Task 1 completed"

async def task2():
    await asyncio.sleep(1)
    raise ValueError("Something went wrong!")

async def main():
    try:
        results = await asyncio.gather(task1(), task2())
        print(results)
    except Exception as e:
        print("Error:", e)

await main()

Error: Something went wrong!


**Using `return_exceptions=True`**

If you want all coroutines to run to completion, even if one fails, use `return_exceptions=True`.

In [55]:
import asyncio

async def task1():
    return "Task 1 completed"

async def task2():
    raise ValueError("Something went wrong!")

async def main():
    results = await asyncio.gather(
        task1(),
        task2(),
        return_exceptions=True
    )
    print(results)

await main()

['Task 1 completed', ValueError('Something went wrong!')]


Instead of raising the exception, it is returned as an element in the results list.

**Summary**

|Default (return_exceptions=False)|	return_exceptions=True|
|---------------------------------|-----------------------|
|First exception is raised immediately.|	Exceptions are returned as results.|
|Other pending coroutines may be cancelled.	|All coroutines are allowed to finish.|
|Use when any failure should stop the operation.	|Use when you want results from all tasks, even if some fail.

**10. What is asyncio.create_task()?**

**Definition: **

`asyncio.create_task()` schedules a coroutine for concurrent execution and returns a `Task` object that can be awaited later.

**Syntax**

```
task = asyncio.create_task(coroutine())
```

**Example**

In [61]:
import asyncio

async def task():
    print("Task started")
    await asyncio.sleep(2)
    print("Task completed")

async def main():
    t = asyncio.create_task(task())  # Start task in the background
    print("Doing other work...")
    await t                          # Wait for the task to finish

await main()

Doing other work...
Task started
Task completed


**Why Use `asyncio.create_task()`?**

- Runs coroutines concurrently.
- Allows other code to execute while the task is running.
- Returns a Task object that you can:
    - await
    - cancel with task.cancel()
    - check with task.done()
    - retrieve the result using task.result() (after completion)

**`await` vs `asyncio.create_task()`**

| `await`                                              | `asyncio.create_task()`                                             |
| ---------------------------------------------------- | ------------------------------------------------------------------- |
| Waits for the coroutine to finish before continuing. | Starts the coroutine immediately and lets it run in the background. |
| Sequential execution.                                | Concurrent execution.                                               |
| Returns the coroutine's result directly.             | Returns a `Task` object that can be awaited later.                  |


**Example Comparison**

In [ ]:
# Sequential
await task1()
await task2()

In [ ]:
# Concurrent
t1 = asyncio.create_task(task1())
t2 = asyncio.create_task(task2())

await t1
await t2

In the second example, `task1()` and `task2()` run concurrently, reducing the total execution time if they are independent.

## **Hard**

**11. Build an async web scraper.**

**Async Web Scraper in Python**

The following example uses aiohttp for asynchronous HTTP requests and BeautifulSoup for HTML parsing. It fetches multiple web pages concurrently and extracts their titles.

Step 1: Install Required Libraries

```
pip install aiohttp beautifulsoup4
```

In [74]:
import asyncio
import aiohttp
from bs4 import BeautifulSoup

# Fetch HTML from a URL
async def fetch(session, url):
    try:
        async with session.get(url, timeout=10) as response:
            response.raise_for_status()
            return await response.text()
    except Exception as e:
        print(f"Error fetching {url}: {e}")
        return None

# Extract page title
def extract_title(html):
    soup = BeautifulSoup(html, "html.parser")
    return soup.title.string.strip() if soup.title else "No Title"

# Process one URL
async def scrape(session, url):
    html = await fetch(session, url)
    if html:
        title = extract_title(html)
        print(f"{url}\nTitle: {title}\n")

# Main function
async def main():
    urls = [
        "https://example.com",
        "https://python.org",
        "https://www.wikipedia.org"
    ]

    async with aiohttp.ClientSession() as session:
        tasks = [asyncio.create_task(scrape(session, url)) for url in urls]
        await asyncio.gather(*tasks)

if __name__ == "__main__":
    await main()

https://example.com
Title: Example Domain

https://python.org
Title: Welcome to Python.org

Error fetching https://www.wikipedia.org: 403, message='Forbidden', url='https://www.wikipedia.org'


**How It Works**

1. `ClientSession()` creates a reusable HTTP session.
2. `fetch()` downloads a page asynchronously.
3. `scrape()` extracts the page title using BeautifulSoup.
4. `asyncio.create_task()` schedules all scraping tasks concurrently.
5. `asyncio.gather()` waits for all tasks to complete.
   
**Advantages**
- Downloads multiple pages concurrently.
- Faster than a synchronous scraper for many URLs.
- Reuses a single HTTP session for efficiency.
- Includes basic error handling for failed requests.

This pattern can be extended to scrape links, headings, tables, images, or other HTML elements, while respecting website terms of service and `robots.txt`.

**12. Download 100 URLs concurrently.***

Below is a scalable example that downloads 100 URLs concurrently using asyncio and aiohttp.

**Install the Required Library**

```
pip install aiohttp
```

In [79]:
import asyncio
import aiohttp

# Download a single URL
async def fetch(session, url):
    try:
        async with session.get(url, timeout=10) as response:
            response.raise_for_status()
            content = await response.text()
            print(f"Downloaded: {url} ({len(content)} bytes)")
            return content
    except Exception as e:
        print(f"Failed: {url} - {e}")
        return None

# Download all URLs concurrently
async def download_all(urls):
    async with aiohttp.ClientSession() as session:
        tasks = [asyncio.create_task(fetch(session, url)) for url in urls]
        results = await asyncio.gather(*tasks, return_exceptions=True)
        return results

async def main():
    # Example: 100 URLs (replace with your own)
    urls = ["https://example.com"] * 10 #modify the number

    results = await download_all(urls)

    successful = sum(r is not None and not isinstance(r, Exception) for r in results)
    print(f"\nSuccessfully downloaded {successful} out of {len(urls)} URLs.")

if __name__ == "__main__":
    await main()

Downloaded: https://example.com (559 bytes)
Downloaded: https://example.com (559 bytes)
Downloaded: https://example.com (559 bytes)
Downloaded: https://example.com (559 bytes)
Downloaded: https://example.com (559 bytes)
Downloaded: https://example.com (559 bytes)
Downloaded: https://example.com (559 bytes)
Downloaded: https://example.com (559 bytes)
Downloaded: https://example.com (559 bytes)
Downloaded: https://example.com (559 bytes)

Successfully downloaded 10 out of 10 URLs.


**Limiting Concurrency (Recommended)**

Launching 100 requests at exactly the same time can overwhelm your system or the target server. A `semaphore` limits the number of concurrent requests.

In [84]:
import asyncio
import aiohttp

CONCURRENT_REQUESTS = 10

async def fetch(session, url, semaphore):
    async with semaphore:
        async with session.get(url) as response:
            return await response.text()

async def main():
    urls = ["https://example.com"] * 100
    semaphore = asyncio.Semaphore(CONCURRENT_REQUESTS)

    async with aiohttp.ClientSession() as session:
        tasks = [
            fetch(session, url, semaphore)
            for url in urls
        ]
        pages = await asyncio.gather(*tasks)

    print(f"Downloaded {len(pages)} pages")

await main()

Downloaded 100 pages


**Why use a semaphore?**

- Prevents sending too many requests at once.
- Reduces the chance of server rate limiting or blocking.
- Lowers memory and network usage.
- Makes the scraper more reliable.

For real-world scraping, limiting concurrency (for example, 10–20 concurrent requests) is generally better than starting all 100 requests simultaneously.

**13. Limiting Concurrency Using `asyncio.Semaphore`**

An `asyncio.Semaphore` limits the number of coroutines that can access a resource at the same time. It is commonly used to control the number of concurrent network requests, preventing server overload and avoiding rate limits.

**Example**

In [90]:
import asyncio
import aiohttp

# Maximum number of concurrent requests
MAX_CONCURRENT = 5

async def fetch(session, url, semaphore):
    async with semaphore:  # Acquire semaphore
        print(f"Fetching: {url}")

        async with session.get(url) as response:
            data = await response.text()

        print(f"Completed: {url}")
        return data

async def main():
    urls = [
        "https://example.com",
        "https://python.org",
        "https://www.wikipedia.org",
        "https://httpbin.org/get",
        "https://github.com",
        "https://stackoverflow.com",
        "https://docs.python.org",
        "https://pypi.org",
        "https://realpython.com",
        "https://aiohttp.readthedocs.io"
    ]

    semaphore = asyncio.Semaphore(MAX_CONCURRENT)

    async with aiohttp.ClientSession() as session:
        tasks = [
            asyncio.create_task(fetch(session, url, semaphore))
            for url in urls
        ]

        results = await asyncio.gather(*tasks)

    print(f"\nDownloaded {len(results)} pages.")

if __name__ == "__main__":
    await main()

Fetching: https://example.com
Fetching: https://python.org
Fetching: https://www.wikipedia.org
Fetching: https://httpbin.org/get
Fetching: https://github.com
Completed: https://example.com
Fetching: https://stackoverflow.com
Completed: https://stackoverflow.com
Fetching: https://docs.python.org
Completed: https://python.org
Completed: https://www.wikipedia.org
Fetching: https://pypi.org
Fetching: https://realpython.com
Completed: https://pypi.org
Fetching: https://aiohttp.readthedocs.io
Completed: https://docs.python.org
Completed: https://realpython.com
Completed: https://github.com
Completed: https://aiohttp.readthedocs.io
Completed: https://httpbin.org/get

Downloaded 10 pages.


**How It Works**

1. **Create a semaphore:**
```
semaphore = asyncio.Semaphore(5)
```
Only 5 coroutines can run inside the semaphore at the same time.

2. **Acquire the semaphore:**
```
async with semaphore:
```
If the limit is reached, additional coroutines wait until a permit becomes available.

3. When a coroutine exits the async with block, the semaphore is automatically released, allowing another waiting coroutine to proceed.

**Execution Flow (Example)**

If there are 10 URLs and the semaphore limit is 3:

```
Time 1: URL1 URL2 URL3
Time 2: URL4 URL5 URL6
Time 3: URL7 URL8 URL9
Time 4: URL10
```

Only 3 requests are active at any moment.

**Advantages**

- Prevents overwhelming the target server.
- Helps avoid rate limiting.
- Reduces memory and network usage.
- Makes asynchronous applications more stable and efficient.

**Summary**

`asyncio.Semaphore(n)` limits the number of concurrently running coroutines to n.
It is commonly used in web scraping, API clients, and other network applications to control concurrency and improve reliability.

**14. Retry Failed API Calls Asynchronously (asyncio + aiohttp)**

A common approach is to retry requests that fail due to temporary issues (timeouts, connection errors, or server errors like HTTP 500). You should also use exponential backoff to avoid overwhelming the server.

In [98]:
import asyncio
import aiohttp

MAX_RETRIES = 3
BASE_DELAY = 1  # seconds

async def fetch(session, url):
    for attempt in range(1, MAX_RETRIES + 1):
        try:
            async with session.get(url, timeout=10) as response:
                response.raise_for_status()
                return await response.json()

        except (
            aiohttp.ClientError,
            asyncio.TimeoutError
        ) as e:
            if attempt == MAX_RETRIES:
                print(f"Failed after {MAX_RETRIES} attempts: {url}")
                return None

            delay = BASE_DELAY * (2 ** (attempt - 1))
            print(
                f"Attempt {attempt} failed for {url}. "
                f"Retrying in {delay} seconds..."
            )
            await asyncio.sleep(delay)

async def main():
    urls = [
        "https://httpbin.org/json",
        "https://httpbin.org/status/500",
        "https://httpbin.org/delay/2",
    ]

    async with aiohttp.ClientSession() as session:
        tasks = [fetch(session, url) for url in urls]
        results = await asyncio.gather(*tasks)

    print(results)

if __name__ == "__main__":
    await main()

Attempt 1 failed for https://httpbin.org/status/500. Retrying in 1 seconds...
Attempt 2 failed for https://httpbin.org/status/500. Retrying in 2 seconds...
Failed after 3 attempts: https://httpbin.org/status/500
Attempt 1 failed for https://httpbin.org/delay/2. Retrying in 1 seconds...
Attempt 1 failed for https://httpbin.org/json. Retrying in 1 seconds...
Attempt 2 failed for https://httpbin.org/json. Retrying in 2 seconds...
[{'slideshow': {'author': 'Yours Truly', 'date': 'date of publication', 'slides': [{'title': 'Wake up to WonderWidgets!', 'type': 'all'}, {'items': ['Why <em>WonderWidgets</em> are great', 'Who <em>buys</em> WonderWidgets'], 'title': 'Overview', 'type': 'all'}], 'title': 'Sample Slide Show'}}, None, {'args': {}, 'data': '', 'files': {}, 'form': {}, 'headers': {'Accept': '*/*', 'Accept-Encoding': 'gzip, deflate, br', 'Host': 'httpbin.org', 'User-Agent': 'Python/3.12 aiohttp/3.10.5', 'X-Amzn-Trace-Id': 'Root=1-6a854e38-072ab73f254f47df6cbcc21e'}, 'origin': '106.51.

**How It Works**

1. Makes an asynchronous API request.
2. If the request succeeds, returns the response.
3. If it fails due to a timeout or network error:
    - Waits before retrying.
    - Uses exponential backoff (1s → 2s → 4s).
4. Stops retrying after the maximum number of attempts.

**Retry with Concurrency Limit**

You can combine retries with a semaphore to prevent too many simultaneous retries.

```
semaphore = asyncio.Semaphore(10)

async def fetch(session, url):
    async with semaphore:
        # retry logic here
        ...
```

**Best Practices**

- Retry temporary failures (timeouts, connection errors, HTTP 429, HTTP 500–504).
- Do not retry most client errors (HTTP 400, 401, 403, 404), as they usually won't succeed without changing the request.
- Use exponential backoff (and ideally add random jitter) to avoid retry storms.
- Set a maximum number of retries.
- Honor the server's Retry-After header for HTTP 429 or 503 responses when provided.

**Summary**
- Use asyncio and aiohttp for asynchronous API calls.
- Retry only transient failures.
- Apply exponential backoff (preferably with jitter).
- Limit concurrency with asyncio.Semaphore for reliable, scalable API clients.

**Design an Async Task Queue**

An async task queue is a system that stores tasks and processes them concurrently using worker coroutines. It is useful for handling background jobs such as API calls, file processing, and web scraping.

**Architecture**

```
          +----------------+
          | Task Producer  |
          +----------------+
                  |
                  v
        +--------------------+
        | asyncio.Queue      |
        +--------------------+
            |    |    |
            v    v    v
       Worker1 Worker2 Worker3
            |    |    |
            +----+----+
                 |
                 v
          Process Results
```

In [104]:
import asyncio
import random

async def worker(worker_id, queue):
    while True:
        task = await queue.get()

        if task is None:      # Shutdown signal
            queue.task_done()
            break

        print(f"Worker {worker_id} processing task {task}")

        await asyncio.sleep(random.randint(1, 3))

        print(f"Worker {worker_id} completed task {task}")

        queue.task_done()


async def producer(queue):
    for i in range(1, 21):
        await queue.put(i)
        print(f"Added task {i}")


async def main():
    queue = asyncio.Queue()

    workers = [
        asyncio.create_task(worker(i, queue))
        for i in range(3)
    ]

    await producer(queue)

    await queue.join()

    # Stop workers
    for _ in workers:
        await queue.put(None)

    await asyncio.gather(*workers)

await main()

Added task 1
Added task 2
Added task 3
Added task 4
Added task 5
Added task 6
Added task 7
Added task 8
Added task 9
Added task 10
Added task 11
Added task 12
Added task 13
Added task 14
Added task 15
Added task 16
Added task 17
Added task 18
Added task 19
Added task 20
Worker 0 processing task 1
Worker 1 processing task 2
Worker 2 processing task 3
Worker 0 completed task 1
Worker 0 processing task 4
Worker 1 completed task 2
Worker 1 processing task 5
Worker 2 completed task 3
Worker 2 processing task 6
Worker 0 completed task 4
Worker 0 processing task 7
Worker 1 completed task 5
Worker 1 processing task 8
Worker 0 completed task 7
Worker 0 processing task 9
Worker 2 completed task 6
Worker 2 processing task 10
Worker 0 completed task 9
Worker 0 processing task 11
Worker 1 completed task 8
Worker 1 processing task 12
Worker 2 completed task 10
Worker 2 processing task 13
Worker 0 completed task 11
Worker 0 processing task 14
Worker 1 completed task 12
Worker 1 processing task 15
Wor

**How It Works**

1. The producer adds tasks to an asyncio.Queue.
2. Multiple workers continuously retrieve tasks from the queue.
3. Each worker processes a task asynchronously.
4. queue.task_done() marks a task as complete.
5. queue.join() waits until all queued tasks have been processed.
6. Special None values act as shutdown signals, allowing workers to exit gracefully.

**Advantages**

- Supports concurrent task processing.
- Automatically balances work across workers.
- Prevents busy waiting by blocking workers until tasks are available.
- Easy to scale by increasing the number of worker coroutines.

**Enhancements for Production**
- Retries: Retry failed tasks with exponential backoff.
- Timeouts: Wrap tasks with `asyncio.wait_for()`.
- Priority queue: Use `asyncio.PriorityQueue` for task prioritization.
- Rate limiting: Use `asyncio.Semaphore` to limit concurrent external requests.
- Result collection: Store results in another queue or return them via `asyncio.gather()`.
- Logging and monitoring: Track task execution time, failures, and queue length.
- Persistent queue: Use systems like Redis, RabbitMQ, or Kafka if tasks must survive process restarts.

**Summary**

An async task queue typically consists of:

- An `asyncio.Queue` to hold tasks.
- One or more producer coroutines that add tasks.
- Multiple worker coroutines that process tasks concurrently.
- Graceful shutdown using sentinel values (such as `None`).
- Optional features like retries, rate limiting, priorities, and persistent storage for production use.